src/b64.mli

```diff
@@ -24,23 +26,34 @@
 
     {e Release %%VERSION%% - %%PKG_HOMEPAGE%%} *)
 
-val default_alphabet : string
-(** A 64-character string specifying the regular Base64 alphabet. *)
+type alphabet
 
-val uri_safe_alphabet : string
-(** A 64-character string specifying the URI- and filename-safe Base64
+val default_alphabet : alphabet
+(** A 64-character alphabet specifying the regular Base64 alphabet. *)
+
+val uri_safe_alphabet : alphabet
+(** A 64-character alphabet specifying the URI- and filename-safe Base64
     alphabet. *)
 
-val decode : ?alphabet:string -> string -> string
+val make_alphabet : string -> alphabet
+val length_alphabet : alphabet -> int
+
+val decode : ?alphabet:alphabet -> string -> string
 (** [decode s] decodes the string [s] that is encoded in Base64 format. Will
     leave trailing NULLs on the string, padding it out to a multiple of 3
-    characters. [alphabet] defaults to {!default_alphabet}. @raise Not_found if
-    [s] is not a valid Base64 string. *)
+    characters. [alphabet] defaults to {!default_alphabet}.
 
-val decode_opt : ?alphabet:string -> string -> string option
+    @raise if Invalid_argument [s] is not a valid Base64 string or if we got a
+    wrong-padding. *)
+
+val decode_opt : ?alphabet:alphabet -> string -> string option
 (** Same as [decode], but returns [None] instead of raising. *)
 
-val encode : ?pad:bool -> ?alphabet:string -> string -> string
+val decode_result : ?alphabet:alphabet -> string -> (string, [ `Malformed | `Wrong_padding ]) result
```

Comment:
I think we should break the interface and make `decode_result` the default, and `decode_exn`.  It seems unnecessary to have `decode_opt` since the result version should be sufficient.

Reply:
we could then even name `decode_result` just `decode` (and IIUC API will break due to `B64` -> `Base64` rename anyways)

Reply:
Yes lets make `decode_result -> decode` and the `B64 -> Base64` and fix up the revdeps.